src/MessagePack.UnityClient/Assets/Scripts/Tests/ShareTests/DynamicObjectResolverConstructorTest.cs

```diff
@@ -181,6 +181,30 @@ public TestConstructor6(object x)
             }
         }
 
+        /// <summary>
+        /// This constructor tests the case where key names differ from property
+        /// names, to ensure that the correct ctor is still found using the member name.
+        /// </summary>
+        [MessagePackObject]
+        public struct TestConstructor7
+        {
+            [Key("x_val")]
+            public int X { get; set; }
+
+            [Key("y_val")]
+            public int Y { get; set; }
+
+            [Key("z_val")]
+            public int Z { get; set; }
+
+            public TestConstructor7(int x, int y, int z)
```

Comment:
Before your change, this ctor's signature would have to be defined as:
```cs
public TestConstructor7(int x_val, int y_val, int z_val)
```
correct?

Please add a test (or point me to an existing one) that uses the constructor signature that works before your change to prove you don't break pre-existing constructors.

Reply:
Hi @AArnott ! You make a good point. I was tempted to point to the cases which use `TestConstructor1`:

https://github.com/neuecc/MessagePack-CSharp/blob/686fcf67690d309a4dd8032260f5616aea6dde5c/src/MessagePack.UnityClient/Assets/Scripts/Tests/ShareTests/DynamicObjectResolverConstructorTest.cs#L11-L48

Since the use of `MessagePackObject(true)` makes the string keys equivalent to the member names, and the ctor parameters are lowercased versions of the member names in that ctor. However, I think it's not explicit enough, so I wrote another test case which should prove that existing behavior is preserved.

https://github.com/neuecc/MessagePack-CSharp/blob/5b30b113e6d73d0396d0a814ecbe39e731dd5b6d/src/MessagePack.UnityClient/Assets/Scripts/Tests/ShareTests/DynamicObjectResolverConstructorTest.cs#L204-L209

Is overloaded into `TestConstructor7` and has ctor parameters that match the explicitly defined string keys on the members. Test case:

https://github.com/neuecc/MessagePack-CSharp/blob/5b30b113e6d73d0396d0a814ecbe39e731dd5b6d/src/MessagePack.UnityClient/Assets/Scripts/Tests/ShareTests/DynamicObjectResolverConstructorTest.cs#L293-L303

passing shows that projects which followed this previously necessary convention will not break.

Let me know if there is anything else you would like me to take a look at! I'm very excited because this is my first time contributing to an open source project, so I am learning a lot, lol.

Nabeel

Reply:
You can't add two constructors to the same type to verify both old and new behavior. MessagePack will choose one deserializer and use it (and it's choosing the one that defines all properties -- not just 2 of them). So you're not actually testing the "old way" right now.

Please create a whole new type with its own constructor that uses the old-style parameter names and a test to exercise it.

Reply:
Hi. I did not know that. It did really throw me for a loop when I added another test case as you recommended and.... it failed.

I think I found a more correct solution however. I rewrote my changes so that now the ctor lookup part checks both string key names and member names. If both are found, it prefers the string key names. I also added a new test case which checks that this more resilient procedure will work even if there is a mix of parameter names matching string key names and parameter names matching member names.

https://github.com/neuecc/MessagePack-CSharp/blob/8ec03d7974d71832694b872b68e40149ad7be9e8/src/MessagePack.UnityClient/Assets/Scripts/Tests/ShareTests/DynamicObjectResolverConstructorTest.cs#L343-L352

So the solution was definitely not as easy as I thought it was, but I think this really should resolve the issue, hopefully. Thanks for giving me the opportunity to learn and figure this out.

Also, as a sidenote, I rebased the last three commits into one commit and pushed that, but I don't know if that was the right thing to do. Maybe it would have been better to wait 🥴

Reply:
Rebasing/squashing is fine IMO during a PR, provided no one else has built on your changes elsewhere, which I expect no one has yet. Thank you for working through this!